includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -174,7 +174,7 @@ public function render( $args, $url ) {
 	 * @return integer|false Video ID, or false if none could be retrieved.
 	 */
 	private function get_video_id_from_url( $url ) {
-		if ( preg_match( '/(?:watch\?v=|embed\/|youtu.be\/)(?P<id>\w*)/', $url, $match ) ) {
+		if ( preg_match( '/(?:watch\?v=|embed\/|youtu.be\/)(?P<id>[\w_-]*)/', $url, $match ) ) {
```

Comment:
Pattern used in 1.4:

https://github.com/ampproject/amp-wp/blob/dce4fae2b4b95008bc60dbd76d1fe603f2aee1ce/includes/embeds/class-amp-youtube-embed.php#L15-L16

Is all of this accounted for?

Reply:
I wonder if the complexity of the pattern is part of the problem for why we missed this. YouTube doesn't actually specify the allowed characters in the video ID, so limiting it to `a-zA-Z0-9_-` may or may not account for everything.

I think we could make it more robust by parsing the URL first:

```php
$parsed_url = wp_parse_url( $url );
$query_vars = [];
if ( isset( $parsed_url['query'] ) ) {
    wp_parse_str( $parsed_url['query'], $query_vars );
    if ( isset( $query_vars['v'] ) ) {
        return $query_vars['v']; // e.g. https://www.youtube.com/watch?v=CMrv_D78oxY
    } elseif ( isset( $query_vars['vi'] ) ) {
        return $query_vars['vi']; // e.g. http://youtube.com/watch?vi=oTJRivZTMLs&feature=channel
    }
}
if ( isset( $parsed_url['path'] ) ) {
    $segments = explode( '/', trim( $parsed_url['path'], '/' ) );
    if ( 1 === count( $segments ) ) {
        return $segments[0]; // e.g. https://youtu.be/XOY3ZUO6P0k
    } elseif ( count( $segments ) > 1 ) {
        if ( in_array( $segments[0], [ 'embed', 'v', 'e', 'vi'], true ) ) {
            return $segments[1]; // e.g. http://www.youtube.com/v/-wtIMTCHWuI?version=3&autohide=1
        }
    }
}
return false;
```

Reply:
Here's a list of a bunch of URLs which could be all accounted for: https://gist.github.com/rodrigoborgesdeoliveira/987683cfbfcc8d800192da1e73adc486

The `www.youtube-nocookie.com` domain also needs to be accounted for.

Not all of those URLs are valid.

Reply:
OK, thanks. Your snippet is committed here verbatim, other than an extra empty line: https://github.com/ampproject/amp-wp/compare/b1344d537c456a926553396d2d90cc39c236f64b...b8d38facdcfb9e10c4c46cbb5cf1b79eaae71fee#diff-23071a193ac2ab7d9be3c9eb4bdebac6R228

There's also [a test](https://github.com/ampproject/amp-wp/compare/b1344d537c456a926553396d2d90cc39c236f64b...b8d38facdcfb9e10c4c46cbb5cf1b79eaae71fee#diff-23071a193ac2ab7d9be3c9eb4bdebac6R252) for `www.youtube-nocookie.com`.